### DISCUSSION TITLE
Issues with types when compiling DAFoam in reverse mode from CodiPack

### DISCUSSION THREAD

**Discussion by uttamcadambi07** [mdolab/dafoam#425] [General]
Hi,

I have a list of lists defined by
```
List<List<scalar>> points;
```
For now let us assume that "something" is indices of a mesh.

Now the length of each of the sublists is different. Somewhere in the code, I collect all the elements that belong to a sublist with index idx and append it to the list of lists "points" in this fashion

```
forAll(points, idx)
{
    scalarList tmp = points[idx];
    forAll(tmp, idx1)
    {
        // do something here that results in "something"
        tmp.append("something");
    }
    points.append(tmp);
    tmp.clear();
}
```

Somewhere in the code, I need to access the sublist of the mainlist ```points``` and i do it this way

```
forAll(points, idx)
{
    scalarList temp = points[idx];
    forAll(tmp, idx1)
    {
        label tempTemp = temp[idx1];
        fieldValue[tempTemp] = a + b;
    }

}
```

Ignore ```a``` and ```b``` they are irrelevant. What is relevant is ```fieldValue[tempTemp]``` where i try to assign a value to the index that is stored in the list of lists points by doing ```fieldValue[points[idx][idx1]]```. The normal version of DAFoam compiles fine, but in the reverse mode, I get this error

```

DAObjFunc/DAMyObjFunc.C: In member function ‘virtual void Foam::DAMyObjFunc::calcObjFunc(const labelList&, const labelList&, Foam::scalarList&, Foam::scalarList&, Foam::scalar&)’:
DAObjFunc/DAMyObjFunc.C:348:39: error: cannot convert ‘codi::ActiveReal<codi::JacobiTape<codi::JacobiTapeTypes<codi::ReverseTapeTypes<double, double, codi::LinearIndexHandler<int> >, codi::ChunkVector> > >’ to ‘Foam::label’ {aka ‘int’} in initialization
  348 |             label tempTemp = temp[idx1];

```

Is there no way to deal with 2D lists with codipack?

**Comment by friedenhe** [comment]
points is a scalarListList, so points[idx][idx1] will return a scalar, you can not use the returned scalar as an index for fieldValue. The original OpenFOAM will not check this (and will probably automatically convert the scalar to a label), but the AD version is more strict and will return an error.

**Reply by uttamcadambi07** [reply]
Ah yes! Massively helpful. Thanks! I fixed it :)
